### PR TITLE
Populate information_schema.key_column_usage table

### PR DESCRIPTION
**Pull request by mxm**
This populates the previously empty table table key_column_usage inside
information_schema with the primary key information of all user tables.

### CODE REVIEW

File: sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java [hunk 1]
```diff
@@ -59,27 +63,28 @@
...
         static final Reference TABLE_NAME = createRef(Columns.TABLE_NAME, DataTypes.STRING);
         static final Reference COLUMN_NAME = createRef(Columns.COLUMN_NAME, DataTypes.STRING);
-        static final Reference ORDINAL_POSITION = createRef(Columns.ORDINAL_POSITION, DataTypes.SHORT);
+        static final Reference ORDINAL_POSITION = createRef(Columns.ORDINAL_POSITION, DataTypes.INTEGER);
     }
 
-    public static ImmutableMap<ColumnIdent, RowCollectExpressionFactory<Void>> expressions() {
-        return ImmutableMap.<ColumnIdent, RowCollectExpressionFactory<Void>>builder()
+    public static ImmutableMap<ColumnIdent, RowCollectExpressionFactory<InformationSchemaIterables.KeyColumnUsage>> expressions() {
+        //noinspection Convert2MethodRef
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java]: Why aren't you using method refs?

File: sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java [hunk 2]
```diff
@@ -59,27 +63,28 @@
...
+        return ImmutableMap.<ColumnIdent, RowCollectExpressionFactory<InformationSchemaIterables.KeyColumnUsage>>builder()
             .put(Columns.CONSTRAINT_CATALOG,
-                () -> RowContextCollectorExpression.objToBytesRef(r -> null))
+                () -> RowContextCollectorExpression.objToBytesRef(k -> k.getSchema()))
             .put(Columns.CONSTRAINT_SCHEMA,
-                () -> RowContextCollectorExpression.objToBytesRef(r -> null))
+                () -> RowContextCollectorExpression.objToBytesRef(k -> "public"))
             .put(Columns.CONSTRAINT_NAME,
-                () -> RowContextCollectorExpression.objToBytesRef(r -> null))
+                () -> RowContextCollectorExpression.objToBytesRef(k -> k.getPkColumnName() + "_pkey"))
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java]: can we ensure that the constraint_name matches the name in `information_schema.table_constraints` ?

File: sql/src/main/java/io/crate/operation/collect/sources/InformationSchemaIterables.java [hunk 3]
```diff
@@ -133,8 +142,18 @@ public InformationSchemaIterables(final Schemas schemas,
         return ingestionRules;
     }
 
-    public Iterable<Void> keyColumnUsageInfos() {
-        return keyColumnUsages;
+    public List<KeyColumnUsage> keyColumnUsage() {
+        return primaryKeyTableInfos.stream()
+            .filter(tableInfo -> !tableInfo.ident().schema().equals(InformationSchemaInfo.NAME))
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/collect/sources/InformationSchemaIterables.java]: I think we should avoid creating an extra list here, it is unnecessary and can easily be avoided ( with `return () -> ...stream()...iterator()`) 

Another way to slightly optimize this might be to have a set with all 3 schema names and then just use 1 filter with a contains check

File: sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java [hunk 4]
```diff
@@ -59,27 +63,28 @@
         static final Reference TABLE_SCHEMA = createRef(Columns.TABLE_SCHEMA, DataTypes.STRING);
         static final Reference TABLE_NAME = createRef(Columns.TABLE_NAME, DataTypes.STRING);
         static final Reference COLUMN_NAME = createRef(Columns.COLUMN_NAME, DataTypes.STRING);
-        static final Reference ORDINAL_POSITION = createRef(Columns.ORDINAL_POSITION, DataTypes.SHORT);
+        static final Reference ORDINAL_POSITION = createRef(Columns.ORDINAL_POSITION, DataTypes.INTEGER);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java]: why the change here? are the column-types of the information schema part of the SQL standard?

Comment (mxm) [reply sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java]: I was looking at https://docs.microsoft.com/en-us/sql/relational-databases/system-information-schema-views/key-column-usage-transact-sql which uses INTEGER.

File: sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java (hunk 1 above)
Comment (mxm) [reply sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java]: Found that they blow up line length here. I find the IntelliJ suggestions a bit too intrusive for this case.

File: sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java (hunk 2 above)
Comment (mxm) [reply sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java]: Good idea!

File: sql/src/main/java/io/crate/operation/collect/sources/InformationSchemaIterables.java (hunk 3 above)
Comment (mxm) [reply sql/src/main/java/io/crate/operation/collect/sources/InformationSchemaIterables.java]: Thanks for the suggestion with the list. I think filter advice is a case for premature optimization.

File: sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java (hunk 1 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java]: In this case I actually agree with Intellij: There is no downside to method refs (well, other than the line length), keeps the code style consistent and it's imho nice for reviews: Method ref syntax can be skipped over more easily. While a lambda might involve closures and other stuff - so I've to read them more carefully

File: sql/src/main/java/io/crate/operation/collect/sources/InformationSchemaIterables.java (hunk 3 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/operation/collect/sources/InformationSchemaIterables.java]: > I think filter advice is a case for premature optimization.

Well, I think it would also be more straightforward way to do that filter anyway =)
Btw, what about `pg_catalog` ?
 I think what's also bugging me about this here is that it's unclear why these 3 schemas are excluded. 
E.g. a `SYS_SCHEMAS.contains..` + expclicit BlobSchema check would convey better why they're excluded.
Would probably be also more resilient towards changes (what if we add another system schema in the future - how would we notice that this needs to be updated)

Comment (mxm) [reply sql/src/main/java/io/crate/operation/collect/sources/InformationSchemaIterables.java]: Ok granted, it is more efficient and keeps the logic in place. Changed it. 

>Would probably be also more resilient towards changes (what if we add another system schema in the future - how would we notice that this needs to be updated)

We would notice immediately because there is a test case which would fail due to new results being returned.

`pg_catalog` does not appear here? https://www.postgresql.org/docs/8.4/static/infoschema-key-column-usage.html What do you mean?

Comment (mfussenegger) [reply sql/src/main/java/io/crate/operation/collect/sources/InformationSchemaIterables.java]: > pg_catalog does not appear here? 

It's another system schema that we have. But I guess there are no tables with primary keys yet.

So I guess this is good as it is if we have a test that starts failing once we add stuff

File: sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java (hunk 1 above)
Comment (mxm) [reply sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java]: ok, i don't mind :)

File: sql/src/main/java/io/crate/operation/collect/sources/InformationSchemaIterables.java (hunk 3 above)
Comment (mxm) [reply sql/src/main/java/io/crate/operation/collect/sources/InformationSchemaIterables.java]: I've added it, just in case.

File: sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java (hunk 4 above)
Comment (mxm) [reply sql/src/main/java/io/crate/metadata/information/InformationKeyColumnUsageTableInfo.java]: The ordinal is a `cardinal_number`: https://www.postgresql.org/docs/8.4/static/infoschema-key-column-usage.html

http://www.npgsql.org/dev/types.html at the bottom:

>        cardinal_number: int4 (must be nonnegative or null)

**Comment by mxm** [comment]
Thanks for the review. Merging with #6384.
